lib/datadog/kit/events.rb

```diff
@@ -7,41 +7,63 @@ module Datadog
       #
       # @param trace [TraceOperation] Trace to attach data to.
-      # @param success [bool] Whether login operation was successful or not.
       # @param user [Hash<Symbol, String>] User information to pass to
       #   Datadog::Kit::Identity.set_user. Must contain at least :id as key.
       # @param others [Hash<String || Symbol, String>] Additional free-form
       #   event information to attach to the trace.
-      def self.track_login(trace, success:, user:, **others)
-        subtag = success ? 'success' : 'failure'
-        event = "users.login.#{subtag}"
+      def self.track_login_success(trace, user:, **others)
+        event = 'users.login.success'
 
-        track(event, trace, **others)
+        track(:appsec, event, trace, **others)
 
         user_options = user.dup
         user_id = user.delete(:id)
 
         raise ArgumentError, 'missing required key: :user => { :id }' if user_id.nil?
 
-        if success
-          Kit::Identity.set_user(trace, id: user_id, **user_options)
-        else
-          trace.set_tag("appsec.events.users.login.#{subtag}.usr.id", user_id)
-        end
+        Kit::Identity.set_user(trace, id: user_id, **user_options)
+      end
+
+      # Attach login failure event information to the trace
+      #
+      # This method is experimental and may change in the future.
+      #
+      # @param trace [TraceOperation] Trace to attach data to.
+      # @param user_id [String] User id that attempted login
+      #   Datadog::Kit::Identity.set_user. Must contain at least :id as key.
+      # @param user_exists [bool] Whether the user id that did a login attempt exists.
+      # @param others [Hash<String || Symbol, String>] Additional free-form
+      #   event information to attach to the trace.
+      def self.track_login_failure(trace, user_id:, user_exists:, **others)
+        event = 'users.login.failure'
+
+        track(:appsec, event, trace, **others)
+
+        raise ArgumentError, 'missing required key: :user => { :id }' if user_id.nil?
+
+        trace.set_tag('appsec.events.users.login.failure.usr.id', user_id)
+        trace.set_tag('appsec.events.users.login.failure.usr.exists', user_exists)
       end
 
       # Attach custom event information to the trace
       #
+      # This method is experimental and may change in the future.
+      #
+      # @param namespace [Symbol] Mandatory. Event namespace. Only :appsec is supported.
       # @param event [String] Mandatory. Event code.
       # @param trace [TraceOperation] Trace to attach data to.
       # @param others [Hash<Symbol, String>] Additional free-form
       #   event information to attach to the trace. Key must not
       #   be :track.
-      def self.track(event, trace, **others)
-        trace.set_tag("appsec.events.#{event}.track", 'true')
+      def self.track(namespace, event, trace, **others)
+        raise ArgumentError, "namespace cannot be #{namespace.inspect}" if namespace.to_sym != :appsec
```

Comment:
By reading the RFC for the track API https://datadoghq.atlassian.net/wiki/spaces/APS/pages/2755793809/Application+Security+Events+Tracking+API+SDK

I can not help to wonder if we need to raise an exception if the namespace is not `appsec`. Do we think other namespaces would use this API outside of `appsec`?

Reply:
There's a use case where customers used that for business events and complained that disabling AppSec also disabled these events, which they still wanted to be reported even though they may not be processed by AppSec backend-side, and that ended up being troublesome for them.

The reverse was true too, as some events must be disabled when AppSec is disabled.

These use cases don't appear to be quite defined yet, so I wanted to keep room for that while being explicit, with the following consequence:

- Anything appsec will be disabled when appsec is disabled
- Anything not appsec will stay enabled when appsec is disabled, whether it ends up being used or not (it sets user-visible tags, and users can do things backend side with user visible tags outside of appsec as they see fit, like filters and triggers and whatnot)

Granted, the current code prevents the second case, but it does so explicitly while leaving room for that to change.

Reply:
I understand. I like the idea of leaving things open for the second use case to be introduce.

Thanks for the detailed explanation